cmd/nerdctl/compose_stop_linux_test.go

```diff
@@ -64,23 +63,36 @@ volumes:
 	projectName := comp.ProjectName()
 	t.Logf("projectName=%q", projectName)
 
+	upAssertHandler := func(svc string) func(stdout string) error {
+		return func(stdout string) error {
+			// Docker Compose v1: "Up", v2: "running"
+			if !strings.Contains(stdout, "Up") && !strings.Contains(stdout, "running") {
+				return fmt.Errorf("service \"%s\" must have been still running", svc)
+			}
+			return nil
+		}
+	}
+
+	exitAssertHandler := func(svc string) func(stdout string) error {
+		return func(stdout string) error {
+			// Docker Compose v1: "Exit code", v2: "exited (code)"
+			if !strings.Contains(stdout, "Exit") && !strings.Contains(stdout, "exited") {
+				return fmt.Errorf("service \"%s\" must have exited", svc)
+			}
+			return nil
+		}
+	}
+
 	base.ComposeCmd("-f", comp.YAMLFullPath(), "up", "-d").AssertOK()
 	defer base.ComposeCmd("-f", comp.YAMLFullPath(), "down", "-v").Run()
 
+	// stop should (only) stop the given service.
 	base.ComposeCmd("-f", comp.YAMLFullPath(), "stop", "db").AssertOK()
-	time.Sleep(3 * time.Second)
-	base.ComposeCmd("-f", comp.YAMLFullPath(), "ps", "db").AssertOutWithFunc(func(stdout string) error {
-		// Docker Compose v1: "Exit code", v2: "exited (code)"
-		if !strings.Contains(stdout, "Exit") && !strings.Contains(stdout, "exited") {
-			return fmt.Errorf("service \"db\" must have exited")
-		}
-		return nil
-	})
-	base.ComposeCmd("-f", comp.YAMLFullPath(), "ps", "wordpress").AssertOutWithFunc(func(stdout string) error {
-		// Docker Compose v1: "Up", v2: "running"
-		if !strings.Contains(stdout, "Up") && !strings.Contains(stdout, "running") {
-			return fmt.Errorf("service \"wordpress\" must have been still running")
-		}
-		return nil
-	})
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "ps", "db").AssertOutWithFunc(exitAssertHandler("db"))
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "ps", "wordpress").AssertOutWithFunc(upAssertHandler("wordpress"))
+
+	// `--timeout` arg should work properly.
+	base.ComposeCmd("-f", comp.YAMLFullPath(), "stop", "-t", "5", "wordpress").AssertOK()
```

Comment:
Does this also check if stop happens after 5 seconds if the processs runs for more than 5 secs?

Reply:
The timeout is for grace shutdown period. At first the command sends a SIGTERM, and wait for (at most) 5 sec (the timeout) before sending a SIGKILL. So in either case the container/service should be stopped. We should be good by checking the expected status after command returns.

I believe the timeout arg only ensures it sends SIGKILL after waiting for the period. It cannot ensure the container is indeed stopped after the period. Sometimes it might take a little longer so the SIGKILL takes effect.

Reply:
Thanks for the input. I'm not sure if that can be ensured. I tested on `docker compose` and it's common to take longer than the time given in `--timeout`. Take the 1st container as an example, I suspect the `8.3s ~= 5s(timeout) + 3.3s(send SIGKILL, SIGKILL take effect, new status returned to the cli)`.

```
$ d compose stop -t 5
[+] Running 24/24
 ⠿ Container compose-test-svcc7-1   Stopped               8.3s
 ⠿ Container compose-test-svccc2-1  Stopped               6.8s
 ⠿ Container compose-test-svc5-1    Stopped               7.6s
 ⠿ Container compose-test-svccc1-1  Stopped               8.4s
 ⠿ Container compose-test-svccc5-1  Stopped               6.9s
 ⠿ Container compose-test-svcc3-1   Stopped               7.1s
 ⠿ Container compose-test-svc4-1    Stopped               7.7s
 ⠿ Container compose-test-svccc3-1  Stopped               7.3s
 ⠿ Container compose-test-svc2-1    Stopped               7.8s
....
```

Probably we can run the command in a goroutine and let the main goroutine sleep longer than 5s (e.g., 8s) and then check the status. But in uncommon case it might become flaky (or not?) if the latency longer than 8s (as shown above in docker example).

Please lmk if this sounds good. Thanks for any other suggestions.

Reply:
hmmm , I expected that `ensure that container do not exceed 5 sec with running state` is hard to be tested too  and for sure we have to avoid flaky test, we are trying hard to make the CI more stable :) , so SGTM to keep it that way ;)

Reply:
thanks :)